File: frontend/src/lib/features.ts
import { lakh } from './format'

interface FeatureMeta {
  label: string
  /** The raw value, formatted for the ledger's value column. */
  value: (v: number | string) => string
  /** A plain-language statement of what the value means for this applicant. */
  says: (v: number | string) => string
}

const n = (v: number | string) => (typeof v === 'number' ? v : Number(v))
const p0 = (v: number) => `${Math.round(v * 100)}%`

export const FEATURES: Record<string, FeatureMeta> = {
  business_vintage: {
    label: 'Business vintage',
    value: (v) => `${n(v).toFixed(1)} yr`,
    says: (v) => `In business for ${n(v).toFixed(1)} years`,
  },
  monthly_revenue: {
    label: 'Monthly revenue',
    value: (v) => lakh(n(v)),
    says: (v) => `Average monthly revenue of ${lakh(n(v))}`,
  },
  revenue_volatility: {
    label: 'Revenue volatility',
    value: (v) => n(v).toFixed(2),
    says: (v) => `Monthly revenue swings by about ${p0(n(v))} around its average`,
  },
  gst_filing_consistency: {
    label: 'GST filing consistency',
    value: (v) => p0(n(v)),
    says: (v) => `${p0(n(v))} of GST returns filed on time in the last year`,
  },
  debt_to_revenue: {
    label: 'Debt-to-revenue',
    value: (v) => n(v).toFixed(2),
    says: (v) => `Existing debt equals ${p0(n(v))} of annual revenue`,
  },
  inflow_outflow_ratio: {
    label: 'Inflow / outflow',
    value: (v) => `${n(v).toFixed(2)}×`,
    says: (v) =>
      n(v) >= 1
        ? `Money coming in is ${n(v).toFixed(2)}× money going out`
        : `Spending exceeds income: only ${n(v).toFixed(2)}× of outflow comes back in`,
  },
  balance_cover: {
    label: 'Balance cover',
    value: (v) => `${n(v).toFixed(2)} mo`,
    says: (v) => `Average bank balance covers ${n(v).toFixed(1)} months of spending`,
  },
  trade_references: {
    label: 'Trade references',
    value: (v) => `${n(v)}`,
    says: (v) => `${n(v)} active supplier ${n(v) === 1 ? 'reference' : 'references'}`,
  },
  delinquency_events: {
    label: 'Past delinquencies',
    value: (v) => `${n(v)}`,
    says: (v) => (n(v) === 0 ? 'No past repayment delays on record' : `${n(v)} past repayment ${n(v) === 1 ? 'delay' : 'delays'} on record`),
  },
  loan_to_revenue: {
    label: 'Loan-to-revenue',
    value: (v) => n(v).toFixed(2),
    says: (v) => `Amount requested is ${p0(n(v))} of annual revenue`,
  },
  kyc_completeness: {
    label: 'KYC completeness',
    value: (v) => p0(n(v)),
    says: (v) => `KYC documents ${p0(n(v))} complete`,
  },
  digital_txn_volume: {
    label: 'Digital payments',
    value: (v) => `${Math.round(n(v))}/mo`,
    says: (v) => `${Math.round(n(v))} digital (UPI) payments a month`,
  },
  sector: {
    label: 'Sector',
    value: (v) => String(v),
    says: (v) => `Operates in ${String(v).toLowerCase()}`,
  },
}

export const featureMeta = (name: string): FeatureMeta =>
  FEATURES[name] ?? { label: name, value: (v) => String(v), says: (v) => `${name}: ${v}` }

File: ml-service/creditsense_ml/features.py
"""The engineered feature vector shared by the ML service and the core backend.

The backend derives these values from an application; the ML service never sees
names, PAN, GSTIN or documents, only this vector.
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import pandas as pd

SECTORS = ["Manufacturing", "Trading", "Services", "Retail", "Agri-allied", "Hospitality"]


@dataclass(frozen=True)
class Feature:
    name: str
    label: str
    unit: str
    lower: float
    upper: float


NUMERIC_FEATURES: list[Feature] = [
    Feature("business_vintage", "Business vintage", "years", 0.0, 100.0),
    Feature("monthly_revenue", "Monthly revenue", "INR lakh", 0.01, 100_000.0),
    Feature("revenue_volatility", "Revenue volatility", "coefficient of variation", 0.0, 5.0),
    Feature("gst_filing_consistency", "GST filing consistency", "share filed on time", 0.0, 1.0),
    Feature("debt_to_revenue", "Debt-to-revenue", "ratio", 0.0, 50.0),
    Feature("inflow_outflow_ratio", "Inflow/outflow ratio", "ratio", 0.0, 10.0),
    Feature("balance_cover", "Balance cover", "months of outflow", 0.0, 60.0),
    Feature("trade_references", "Trade references", "count", 0.0, 500.0),
    Feature("delinquency_events", "Delinquency events", "count", 0.0, 100.0),
    Feature("loan_to_revenue", "Loan-to-revenue", "ratio", 0.0, 50.0),
    Feature("kyc_completeness", "KYC completeness", "score", 0.0, 1.0),
    Feature("digital_txn_volume", "Digital transaction volume", "transactions per month", 0.0, 1_000_000.0),
]
NUMERIC_NAMES = [f.name for f in NUMERIC_FEATURES]
FEATURE_NAMES = NUMERIC_NAMES + ["sector"]
SECTOR_DUMMIES = [f"sector_{s}" for s in SECTORS[1:]]  # Manufacturing is the reference level
MODEL_COLUMNS = NUMERIC_NAMES + SECTOR_DUMMIES


def to_model_matrix(df: pd.DataFrame) -> pd.DataFrame:
    """One-hot encode the sector and order columns exactly as the models expect."""
    out = df[NUMERIC_NAMES].astype(float).copy()
    for s in SECTORS[1:]:
        out[f"sector_{s}"] = (df["sector"] == s).astype(float)
    return out[MODEL_COLUMNS]


def group_sector(contrib: np.ndarray) -> dict[str, float]:
    """Collapse per-column attributions back to one value per business feature."""
    values = dict(zip(MODEL_COLUMNS, contrib.tolist()))
    grouped = {name: values[name] for name in NUMERIC_NAMES}
    grouped["sector"] = float(sum(values[c] for c in SECTOR_DUMMIES))
    return grouped
